[MRG+1] Script for example ensemble OOB error plot. Fixes #4273

kianho opened the pull request:
This script generates an example plot of the OOB error vs. n_estimators using the `RandomForestClassifier(warm_start=True)` approach, for inclusion in the website docs/examples section, as suggested by @glouppe and @amueller in #4273.

Please advise if further amendments are required.

Commit:
Initial commit of plot_ensemble_oob.py

This is a proposed fix for #4273.

Commit:
Minor amendments to plot_ensemble_oob.py.

Corrected typo in the docstring.

Removed explicit figure size statement.

Fixed docstring formatting.

Corrected the author list.

amueller commented on examples/ensemble/plot_ensemble_oob.py:
```diff
@@ -0,0 +1,89 @@
+#         
+# License: BSD 3 Clause
+
+import pylab
```
Please don't use pylab, instead use matplotlib.pyplot.

amueller commented:
Thanks, I think this is a good example. It would be nice to make it run a little bit faster, though. It currently runs ~30s on my machine.
Also, could you add it to the narrative documentation, maybe as a "see this example" when talking about oob scores or when talking about the number of estimators?

amueller commented:
Also, could you please run pep8 on the example?

amueller commented:
Also, maybe start with n_estimators=5, not 1, which will make the warning disappear (for one estimator, not everything has oob scores)

amueller commented:
If you reduce n_features you can also reduce n_estimators until the curve converges, and you can get down to <10s.

kianho commented:
> Also, could you add it to the narrative documentation (user guide), maybe as a "see this example" 
> when talking about oob scores or when talking about the number of estimators?
I'd be happy to take care of this aswell, looking through the user guide, I couldn't find any mention of OOB in the random forests section, but it is mentioned in the ensemble regularization section (http://scikit-learn.org/stable/modules/ensemble.html#subsampling), I could add it to  add it to the list of examples there?

Also, do you think I should augment this script to include the forest ensemble regressors? 

Thanks for the feedback.

amueller commented:
@kianho I think just doing the classifier is fine. And adding it to the subsampling section as example seems good.
Please remove the 4th commit if you don't want it merged. Just ``git rebase -i HEAD~5`` and delete it.

amueller commented:
Also, please don't merge master into your branch. If you want to update your branch, rebase it on top of master.

amueller commented:
maybe it would also be nice if you squash your commits.

Commit:
Amended plot_ensemble_oob.py according to #4665.

@amueller provided feedback on improving my original PR (#4665) of the
plot_ensemble_oob.py script.

A number of major changes were made accordingly:

- Used `matplotlib.pyplot` instead of `pylab`.

- To improve the run-time to <10secs, I reduced the dimensionality of
  the sample dataset and set the max. number of estimators to 150.

- To avoid OOB warnings, the min. number of estimators was set to 15.
  Values <15 would still raise the warnings.

- The script is PEP8-compliant via the `pep8` command-line script. I
  needed to move `print(__doc__)` and author list comments.

- Re-added @amueller to the author list (had mistakenly been removed).

- Added a link to this example to the user-guide under the `Ensemble
  Methods` section.

kianho commented:
To undo the unwanted merging of master into this branch, I reverted to the commit prior to that merge and re-applied the "major amendment" commit. As far as I can tell, everything is fine (aside from the non-squashing of minor commits).

amueller commented on examples/ensemble/plot_ensemble_oob.py:
```diff
@@ -0,0 +1,91 @@
+OOB Errors for Random Forests and Extra Trees Classifiers
+=========================================================
+
+The ``RandomForestClassifier`` and ``ExtraTreesClasifier`` are trained using
```
Actually, ExtraTrees don't use bootstrap by default.

amueller replied:
Actually, maybe @glouppe wants to comment on this. I'm not sure if people use ExtraTrees with bootstrap=True.
Btw, there is a typo in ExtraTreesClassifier here.

amueller commented on examples/ensemble/plot_ensemble_oob.py:
```diff
@@ -0,0 +1,91 @@
+bootstrap sample of the training observations :math:`z_i = (x_i, y_i)`. The
+*out-of-bag* (OOB) error is the average prediction error for each :math:`z_i`
+from trees that do not contain :math:`z_i` in their respective bootstrap
+sample. This allows models to be simultaneously fit and cross-validated [1].
```
I would say validated, not cross-validated.

kianho replied:
p593 of Hastie et al. states:

> Hence unlike many other nonlinear estimators,
random forests can be fit in one sequence, with cross-validation being per-
formed along the way.

while not a standard form of cross-validation, are the OOB errors technically not estimated via type of cross-validation? However, I'm happy to change to "validation" to avoid potential confusion with something like k-fold cross validation.

amueller commented on examples/ensemble/plot_ensemble_oob.py:
```diff
@@ -0,0 +1,91 @@
+to approximate the error stabilization point of each model at which training
+can be halted.
+
+.. [1] T. Hastie, R. Tibshirani and J. Friedman, "Elements of Statistical
```
Can you give a page number?

amueller commented on examples/ensemble/plot_ensemble_oob.py:
```diff
@@ -0,0 +1,91 @@
+
+The example below demonstrates how the OOB error can be measured at the
+inclusion of each new tree whilst fitting ``RandomForestClassifier`` and
+``ExtraTreesClassifier`` models. The subsequent plot enables the practitioner
```
Remove extra trees, I think.

amueller commented:
As I mentioned, I'm unsure about including extratees. I overlooked that before. Maybe @glouppe or @arjoly can comment on that?

Otherwise this LGTM.

amueller commented on examples/ensemble/plot_ensemble_oob.py:
```diff
@@ -0,0 +1,91 @@
+from sklearn.datasets import make_classification
+from sklearn.ensemble import RandomForestClassifier, ExtraTreesClassifier
+
+# Author: Kian Ho <hui.kian.ho@gmail.com>
```
Btw, feel free to just give yourself as the author, but you can keep us if you like.

arjoly commented:
> As I mentioned, I'm unsure about including extratees. I overlooked that before. Maybe @glouppe or @arjoly can comment on that?

I would remove those, since there is no bootstrap samples.

arjoly commented:
One issue with the current oob_score implementation is that there is no check that the X and the y have not been changed between two call to fit with `warm_start=True`. This might lead to incorrect oob score.

@glouppe What is your opinion in light of the discussion #4500?

arjoly commented:
Can you show the resulting plot in the pull request?

kianho commented:
Here is the plot:

![plot_ensemble_oob](https://cloud.githubusercontent.com/assets/783366/7509358/023793b6-f4d3-11e4-9a81-b989aa85689a.png)

Commit:
Made more amendments according to PR feedback.

A number of further amendments to the plot_ensemble_oob.py example
script were suggested in the PR thread and addressed accordingly:

- The ExtraTreesClassifier models were removed from the example, since
  they don't use bootstrapping by default (but can be using bootstrap=True).

- Included the OOB errors for RandomForestClassifier models with various
  max_features values.

- Changed the sample datasets to make for a nicer looking plot.

- Changed "cross-validated" to "validated" in the docstring.

- Added the relevant page numbers to the Hastie et al. reference.

- PEP8 compliance, fixed line > 80 chars.

amueller commented on examples/ensemble/plot_ensemble_oob.py:
```diff
@@ -0,0 +1,86 @@
+observations :math:`z_i = (x_i, y_i)`. The *out-of-bag* (OOB) error is the
+average prediction error for each :math:`z_i` from trees that do not contain
+:math:`z_i` in their respective bootstrap sample. This allows models to be
+simultaneously fit and validated [1].
```
Thanks for the reference. I guess the cross-validated is more clear in their text as they refer to the exercise that shows the identity. I think you text now is good.

glouppe commented:
This is a nice and helpful example, thanks! Nothing more to comment on my side. +1 for merge

GaelVaroquaux commented:
LGTM. Thanks. Merging.

kianho commented:
Thanks @glouppe @arjoly @amueller for the discussion and feedback. Looking forward to making more contributions in the future.

arjoly commented:
Thanks @kianho
